snuba/state/cache/redis/backend.py

```diff
@@ -0,0 +1,143 @@
+from snuba.state.cache.abstract import Cache, ExecutionError, ExecutionTimeoutError
+from snuba.utils.codecs import Codec
+
+
+T = TypeVar("T")
+
+
+logger = logging.getLogger(__name__)
+
+
+RESULT_VALUE = 0
+RESULT_EXECUTE = 1
+RESULT_WAIT = 2
+
+
+class RedisCache(Cache[T]):
+    def __init__(
+        self, client: RedisClientType, prefix: str, codec: Codec[str, T]
+    ) -> None:
+        self.__client = client
+        self.__prefix = prefix
+        self.__codec = codec
+
+        # TODO: This should probably be lazily instantiated, rather than
+        # automatically happening at startup.
+        self.__script_get = client.register_script(
+            resource_string("snuba", "state/cache/redis/scripts/get.lua")
+        )
+        self.__script_set = client.register_script(
+            resource_string("snuba", "state/cache/redis/scripts/set.lua")
+        )
+
+    def __build_key(self, key: str) -> str:
+        return f"{self.__prefix}{{{key}}}"
+
+    def get(self, key: str) -> Optional[T]:
+        value = self.__client.get(self.__build_key(key))
+        if value is None:
+            return None
+
+        return self.__codec.decode(value)
+
+    def set(self, key: str, value: T) -> None:
+        self.__client.set(
+            self.__build_key(key),
+            self.__codec.encode(value),
+            ex=get_config("cache_expiry_sec", 1),
+        )
+
+    def get_or_execute(self, key: str, function: Callable[[], T], timeout: int) -> T:
+        result_key = self.__build_key(key)
+        wait_queue_key = f"{self.__build_key(key)}:w"
+        task_ident_key = f"{self.__build_key(key)}:t"
+
+        result = self.__script_get(
+            [result_key, wait_queue_key, task_ident_key], [timeout, uuid.uuid1().hex]
+        )
+
+        if result[0] == RESULT_VALUE:
+            logger.debug("Immediately returning result from cache hit.")
+            return self.__codec.decode(result[1])
+        elif result[0] == RESULT_EXECUTE:
+            task_ident = result[1].decode("utf-8")
+            task_timeout = int(result[2])
+            logger.debug(
+                "Executing task (%r) with %s second timeout...",
+                task_ident,
+                task_timeout,
+            )
+
+            argv = [task_ident, 60]
+            try:
+                value = function()
+            except Exception:
+                raise
+            else:
+                argv.extend(
+                    [self.__codec.encode(value), get_config("cache_expiry_sec", 1)]
+                )
+            finally:
+                logger.debug("Setting result and waking blocked clients...")
+                self.__script_set(
+                    [
+                        result_key,
+                        wait_queue_key,
+                        task_ident_key,
+                        self.__build_notify_queue_key(key, task_ident),
+                    ],
+                    argv,
+                )
```

Comment:
This currently throws in the case of a task ident mismatch, this should probably be logged as a warning instead.